Eng/pr getters setters

bro4all opened the pull request:
<!-- What's in this pull request? -->
In this patch, we add the ability to import struct getters/setters 

example:
```
struct X {
    int getX() const {}
    void setX(int) {}
};

// will be:
var Object = X()
Object.x = 1

```

bro4all commented:
@swift-ci please smoke test.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -103,7 +103,11 @@ enum class MakeStructRawValuedFlags {
 using MakeStructRawValuedOptions = OptionSet<MakeStructRawValuedFlags>;
 } // end anonymous namespace
-
+Expr *createSelfExpr(AccessorDecl *accessorDecl);
+DeclRefExpr *createParamRefExpr(AccessorDecl *accessorDecl, unsigned index);
+CallExpr * createAccessorImplCallExpr(FuncDecl *accessorImpl,
+                           Expr *selfExpr,
+                           DeclRefExpr *keyRefExpr = nullptr);
```
Remove these forward declarations.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3479,6 +3483,11 @@ namespace {
           continue;
         }
+        
+        if (isa<VarDecl>(member) && isa<clang::CXXMethodDecl>(nd)) {
```
Please add a comment about this case.

zoecarver commented on lib/ClangImporter/ImporterImpl.h:
```diff
@@ -499,7 +499,7 @@ class LLVM_LIBRARY_VISIBILITY ClangImporter::Implementation
   /// pairs.
   llvm::DenseMap<std::pair<FuncDecl *, FuncDecl *>, SubscriptDecl *> Subscripts;
-
+  llvm::DenseMap<Identifier, std::tuple<VarDecl *, AccessorDecl *, AccessorDecl *>> GetterSetterMap;
```
Please add a comment to this like the one for `Subscripts`.

zoecarver replied:
Also, I wonder if there's a better way to hold the values than a tuple.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3577,7 +3587,9 @@ namespace {
           structResult->setIsCxxNonTrivial(
               !cxxRecordDecl->isTriviallyCopyable());
-
+        for (auto &getterAndSetter : Impl.GetterSetterMap){
```
This should be called `computedProperties` or something.

bro4all replied:
These `VarDecl`s will be added to _every_ record.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3577,7 +3587,9 @@ namespace {
               !cxxRecordDecl->isTriviallyCopyable());
-
+        for (auto &getterAndSetter : Impl.GetterSetterMap){
+            result->addMember(get<0>(getterAndSetter.second));
```
If we can't get rid of the tuple, please add a comment explaining that this is the vardecl/property itself.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4237,9 +4249,170 @@ namespace {
 
       recordObjCOverride(result);
     }
+      static std::pair<BraceStmt *, bool>
```
Please run this whole patch through clang format.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4237,9 +4249,170 @@ namespace {
+    AccessorDecl *tryCreateGetterAccessor(const clang::CXXMethodDecl *clangMeth,
+                                          VarDecl *swiftVar, FuncDecl *method) {
+     
+      // If we have a constexpr var with an evaluated value, try to create an
```
This comment is wrong.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4237,9 +4249,170 @@ namespace {
+      // accessor for it. If we can remove all references to the global (which
+      // we should be able to do for constexprs) then we can remove the global
+      // entirely.
+      auto bodyParams = method->getParameters();
```
This can probably be inlined.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4237,9 +4249,170 @@ namespace {
+
+          Type elementTy = valueParamRefExpr->getDecl()->getInterfaceType();
+
+          auto *getterImplCallExpr =
```
rename

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4237,9 +4249,170 @@ namespace {
+                  Impl.SwiftContext, SourceLoc(), SourceLoc(), AccessorKind::Set,
+                  swiftVar, SourceLoc(), StaticSpellingKind::None,
+                  /*Async=*/false, /*AsyncLoc=*/SourceLoc(),
+                  /*throws=*/false, SourceLoc(), /*genericParams*/ nullptr,bodyParams, swiftVar->getASTContext().getVoidType(),
```
```suggestion
                  /*throws=*/false, SourceLoc(), /*genericParams*/ nullptr,bodyParams, ctx.getVoidType(),
```

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4237,9 +4249,170 @@ namespace {
+
+
+
+      Decl *VisitCXXMethodDecl(const clang::CXXMethodDecl *decl) {
```
I think you can clean up and refactor this whole function.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -7518,7 +7691,7 @@ SwiftDeclConverter::importAccessor(const clang::ObjCMethodDecl *clangAccessor,
   return accessor;
 }
 
-static Expr *createSelfExpr(AccessorDecl *accessorDecl) {
```
Along with the first comment in this review, you can add back the static keyword. (You'll need to move these functions elsewhere, I think. Specifically, move them to before their first use.)

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -7568,7 +7741,14 @@ createAccessorImplCallExpr(FuncDecl *accessorImpl,
 
-  auto *argList = ArgumentList::forImplicitUnlabeled(ctx, {keyRefExpr});
+  ArgumentList *argList;
+  if (keyRefExpr) {
```
Remove.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -7568,7 +7741,14 @@ createAccessorImplCallExpr(FuncDecl *accessorImpl,
   accessorImplDotCallExpr->setThrows(false);
 
-  auto *argList = ArgumentList::forImplicitUnlabeled(ctx, {keyRefExpr});
+  ArgumentList *argList;
```
Nit: please initialize this to `nullptr`.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -8930,6 +9110,18 @@ ClangImporter::Implementation::importDeclImpl(const clang::NamedDecl *ClangDecl,
     Result = importDecl(UnderlyingDecl, version);
     SkippedOverTypedef = true;
   }
+  // TODO extend Swift classes to it's equivalents from it's ClanDecled attributes
```
I don't understand this TODO.

zoecarver reviewed (commented):
Hey Omar! This patch looks great! I'm super excited to start importing these methods more ergonomically! 

Don't be overwhelmed by the comments, the logic actually looks pretty good, so most of these are just nits. Please add some test cases as well. 

Also, I think we may need a forum post for this patch at some point. But we can probably get to that later.

zoecarver commented:
I'd specifically like to see some tests that check for the following cases:

```
void getX();
void get();
int getX();
int get();
int getfoo();
int getFOO();
int GetFoo();
int Getter();
```
And 
```
void setX();
void set();
int setX(int);
void set(int);
...
```
And at least one test when we have a setter but no getter.

egorzhdan reviewed (commented):
This looks great! Other than what Zoe already mentioned, I have a few ideas for test cases:

```cpp
int& getInt() const;
void setInt(int);
// I don't know if these should be transformed into a computed property, I think, for now we should just make sure it doesn't crash :)

int& getInt2() const;
void setInt2(int*);
// Same as above.

void settings(int);
void setUp(int);
// These should not be transformed into a computed property.

int getOrCreateSomething() const;
// This would probably be transformed into something awkward (`var orCreateSomething: Int`), and I think it's fine.
// It would be nice to still be able to call the getters or setters explicitly
// (in this case, `let x = object.getOrCreateSomething()` instead of `let x = object.orCreateSomething`)
// I think it's worth having a test for this.
```

bro4all commented:
Another test case: make sure that it doesn't matter which order getters/setters are defined in (i.e., you can define the setter first).

bro4all commented:
> It would be nice to still be able to call the getters or setters explicitly

Yes, I think we should always continue to import the original method. Maybe we want to add a deprecation, though. I think this happens for some Objective-C enums. Maybe something for the forums to decide.

Commit:
fixing setters

Commit:
adding tests

bro4all commented:
@swift-ci please smoke test.

bro4all commented:
@swift-ci please smoke test

bro4all commented:
@swift-ci please smoke test

hyp commented:
@swift-ci please smoke test

Commit:
Removing extras

Commit:
fixing tests

bro4all commented:
@swift-ci please smoke test

bro4all commented:
@swift-ci please smoke test

Commit:
Better naming for properties

Commit:
Cleanup

Commit:
Add tests

Commit:
Clang format it

Commit:
more refactoring and some more tests

Commit:
More tests and more fixes

Commit:
Updating tests:

fixing other tests to work with new property importing

bro4all commented:
@swift-ci please smoke test

Commit:
Fix the two asserts. Move things around. Remove createImported. Move CXXMethodBridging to it's own header.

Commit:
General updates:

Fixing tests, adding radars to follow issues that cna be a started issues on the project.
Also Factoring out MethodBridging.

bro4all commented:
@swift-ci please smoke test

zoecarver commented on include/swift/ClangImporter/CXXMethodBridging.h:
```diff
@@ -0,0 +1,150 @@
+//
```
I think this needs to be different. Something like:
```
//===--- GenValueWitness.cpp - IR generation for value witnesses ----------===//
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
//
//===----------------------------------------------------------------------===//
//
//  This file implements IR generation for value witnesses in Swift.
//
//  Value witnesses are (predominantly) functions that implement the basic
//  operations for copying and destroying values.
//
//  In the comments throughout this file, three type names are used:
//    'B' is the type of a fixed-size buffer
//    'T' is the type which implements a protocol
//
//===----------------------------------------------------------------------===//
```

zoecarver commented on include/swift/ClangImporter/CXXMethodBridging.h:
```diff
@@ -0,0 +1,150 @@
+
+#include <string>
+
+struct CXXMethodBridging {
```
This should go in the swift namespace

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -3507,6 +3571,9 @@ namespace {
           continue;
         }
 
+        if(nd->getDeclName().isIdentifier())
```
nit: space

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -4308,9 +4404,53 @@ namespace {
 
       recordObjCOverride(result);
     }
```
Nit: new line.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -7765,6 +7848,80 @@ synthesizeSubscriptSetterBody(AbstractFunctionDecl *afd, void *context) {
+      setterDecl->setSelfAccessKind(SelfAccessKind::Mutating);
+      result->setIsSetterMutating(true);
+    } else {
+        setterDecl->setSelfAccessKind(SelfAccessKind::NonMutating);
```
nit: indentation.

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -9027,6 +9184,25 @@ ClangImporter::Implementation::importDeclImpl(const clang::NamedDecl *ClangDecl,
     Result = importDecl(UnderlyingDecl, version);
     SkippedOverTypedef = true;
   }
+  // TODO extend Swift classes to it's equivalents from it's ClanDecled
```
What does this mean?

zoecarver commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -9027,6 +9184,25 @@ ClangImporter::Implementation::importDeclImpl(const clang::NamedDecl *ClangDecl,
     Result = importDecl(UnderlyingDecl, version);
     SkippedOverTypedef = true;
   }
+  // TODO extend Swift classes to it's equivalents from it's ClanDecled
+  // attributes
+
+  if (isa<clang::FunctionDecl>(ClangDecl)) {
+    if (ClangDecl->getDeclName().isIdentifier() &&
+        ClangDecl->getName().startswith_insensitive("get") &&
+        ClangDecl->getName().lower() != "getter") {
+      const_cast<clang::NamedDecl *>(ClangDecl)->addAttr(
+          clang::SwiftAttrAttr::CreateImplicit(ClangDecl->getASTContext(),
+                                               "import_as_getter"));
+    }
+    if (ClangDecl->getDeclName().isIdentifier() &&
+        ClangDecl->getName().startswith_insensitive("set") &&
+        ClangDecl->getName().lower() != "setter") {
+      const_cast<clang::NamedDecl *>(ClangDecl)->addAttr(
+          clang::SwiftAttrAttr::CreateImplicit(ClangDecl->getASTContext(),
+                                               "import_as_setter"));
+    }
+  }
```
This can all be removed.

zoecarver commented on test/Interop/Cxx/ergonomics/Inputs/implicit-computed-properties.h:
```diff
@@ -0,0 +1,189 @@
+#ifndef SWIFT_IMPLICIT_COMPUTED_PROPERTIES_H
+#define SWIFT_IMPLICIT_COMPUTED_PROPERTIES_H
```
Everything here needs to have consistent formatting that matches the rest of the project (sans the names obviously).

zoecarver replied:
I'm really just saying indent everything with two spaces.

zoecarver commented on test/Interop/Cxx/ergonomics/implicit-computed-properties.swift:
```diff
@@ -0,0 +1,70 @@
+    expectEqual(a.foo, 42)
+    a.foo = 32
+    expectEqual(a.foo, 32)
+
```
Space

zoecarver commented on test/Interop/Cxx/ergonomics/implicit-computed-properties.swift:
```diff
@@ -0,0 +1,70 @@
+import StdlibUnittest
+import ImplicitComputedProperties
+
+var ImplicitComputedPropertiesTestSuite = TestSuite("ImplicitComputedProperties")
```
Same comment about formatting.

zoecarver commented on test/Interop/Cxx/ergonomics/implicit-computed-properties.swift:
```diff
@@ -0,0 +1,70 @@
+    expectEqual(e.utf8Str, 32)
+}
+ImplicitComputedPropertiesTestSuite.test("setters") {
+    var f = IntGetterSetter()
```
Can you test the getter first so we know the setter actually did something?

zoecarver commented on test/Interop/Cxx/ergonomics/implicit-computed-properties.swift:
```diff
@@ -0,0 +1,70 @@
+}
+
+ImplicitComputedPropertiesTestSuite.test("non trivial") {
+    var Object = NonTrivialGetterSetter()
```
Same: please test the getter first.

zoecarver commented on test/Interop/Cxx/ergonomics/implicit-computed-properties-module-interface.swift:
```diff
@@ -0,0 +1,258 @@
+// CHECK-NEXT:    init()
+// CHECK-NEXT:    init(val: Int32)
+// CHECK-NEXT:    var x: Int32 { mutating get set }
+// CHECK-NEXT:    mutating func getX() -> Int32
```
Interesting: so these both got imported as mutating. We need a test where these are not imported as mutating.

zoecarver replied:
You can either create a global, or (probably better) just write a class like this with an empty setter:
```
struct ConstGetterAndSetter {
   int getfoo() const { return 42; }
   void setfoo(int v) const { }
 };
```
Still test the setter, but know the value won't change. That's OK.

zoecarver commented on include/swift/ClangImporter/CXXMethodBridging.h:
```diff
@@ -0,0 +1,150 @@
+            return Kind::getter;
+        }
+
+        rdar://89453187 (Add subscripts clarification to CXXMethod Bridging to clean up importDecl)
```
Comment.

zoecarver replied:
```suggestion
        // TODO: classify subscripts here (rdar://89453187).
```

zoecarver commented on include/swift/ClangImporter/CXXMethodBridging.h:
```diff
@@ -0,0 +1,150 @@
+        if (nameIsBlacklist())
+            return Kind::unkown;
+
+        // this should be handled as snake case. See: rdar://89453010
+        // case. In the future we could
+        //  import these too, though.
```
Nit:
```suggestion
        // TODO: this should be handle snake case too (rdar://89453010).
```

zoecarver commented on test/Interop/Cxx/ergonomics/Inputs/implicit-computed-properties.h:
```diff
@@ -0,0 +1,189 @@
+  void setX(int);
+};
+
+struct VoidGetterNoName {
```
VoidSetterNoName

zoecarver commented on test/Interop/Cxx/ergonomics/Inputs/implicit-computed-properties.h:
```diff
@@ -0,0 +1,189 @@
+  void setX(int);
+};
+
+struct VoidGetterNoName {
+  void set();
+};
+
+struct IllegalIntReturnSetter {
+  int setX(int);
+};
+
+struct TwoParameterSetter {
+  void setX(int, int);
+};
+
+struct NoNameSetter {
+  void set(int);
+};
```
These all need correct getters. Otherwise they're all testing the same thing.

zoecarver commented on test/Interop/Cxx/ergonomics/Inputs/implicit-computed-properties.h:
```diff
@@ -0,0 +1,189 @@
+struct NoNameSetter {
+  void set(int);
+};
+
```
Please add a test case for `NoGetter`.

zoecarver commented on test/Interop/Cxx/ergonomics/Inputs/implicit-computed-properties.h:
```diff
@@ -27,16 +27,21 @@ struct NoNameVoidGetter {
+  void setFOO(int v) const {value = v ;}
 };
 
 struct LongNameMix {
```
LongNameMix doesn't really make sense here. This shows the need for two different tests:
```
struct UpperCaseMix {
  int getFoo() const;
  void SetFoo(int);
}
```
And:
```
struct UpperCaseGetterSetter {
  ...
}
```

zoecarver commented:
It's worrying to see any tests passing... this shouldn't compile.

zoecarver commented:
@egorzhdan want to take another look over this?

Commit:
Fixing Comments left on the PR:

General formatting.

Commit:
Fixing tests, and general updates for formatting

zoecarver commented:
@swift-ci please test.

zoecarver commented on include/swift/ClangImporter/CXXMethodBridging.h:
```diff
@@ -0,0 +1,172 @@
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements IR generation for value witnesses in Swift.
+//
+//  Value witnesses are (predominantly) functions that implement the basic
+//  operations for copying and destroying values.
+//
+//  In the comments throughout this file, three type names are used:
+//    'B' is the type of a fixed-size buffer
+//    'T' is the type which implements a protocol
+//
+//===----------------------------------------------------------------------===//
```
Sorry can you delete this or change it?

Commit:
removing extras and passing this on swift.

zoecarver commented:
@swift-ci please test.
